File: win-prog/ProgWinVB/Paths Regions and Clipping/KeyholeBitmap/KeyholeBitmap.vb
'----------------------------------------------
' KeyholeBitmap.vb (c) 2002 by Charles Petzold
'----------------------------------------------
Imports System
Imports System.Drawing
Imports System.Drawing.Drawing2D
Imports System.Drawing.Imaging
Imports System.Windows.Forms

Class KeyholeBitmap
    Inherits PrintableForm

    Private bm As Bitmap

    Shared Shadows Sub Main()
        Application.Run(New KeyholeBitmap())
    End Sub

    Sub New()
        Text = "Keyhole Bitmap"

        ' Load image.
        Dim img As Image = Image.FromFile( _
                    "..\..\..\Images and Bitmaps\Apollo11FullColor.jpg")

        ' Create clipping path.
        Dim path As New GraphicsPath()
        path.AddArc(80, 0, 80, 80, 45, -270)
        path.AddLine(70, 180, 170, 180)

        ' Get size of clipping path.
        Dim rectf As RectangleF = path.GetBounds()

        ' Create new bitmap initialized to transparent.
        bm = New Bitmap(CInt(rectf.Width), CInt(rectf.Height), _
                        PixelFormat.Format32bppArgb)

        ' Create Graphics Object based on new bitmap.
        Dim grfx As Graphics = Graphics.FromImage(bm)

        ' Draw original image on bitmap with clipping.
        grfx.SetClip(path)
        grfx.TranslateClip(-rectf.X, -rectf.Y)
        grfx.DrawImage(img, CInt(-rectf.X), CInt(-rectf.Y), _
                           img.Width, img.Height)
        grfx.Dispose()
    End Sub

    Protected Overrides Sub DoPage(ByVal grfx As Graphics, _
            ByVal clr As Color, ByVal cx As Integer, ByVal cy As Integer)
        grfx.DrawImage(bm, (cx - bm.Width) \ 2, _
                           (cy - bm.Height) \ 2, _
                           bm.Width, bm.Height)
    End Sub
End Class

File: win-prog/ProgWinCS/Paths Regions and Clipping/KeyholeBitmap/KeyholeBitmap.cs
//--------------------------------------------
// KeyholeBitmap.cs © 2001 by Charles Petzold
//--------------------------------------------
using System;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Imaging;
using System.Windows.Forms;

class KeyholeBitmap: PrintableForm
{
     Bitmap bitmap;

     public new static void Main()
     {
          Application.Run(new KeyholeBitmap());
     }
     public KeyholeBitmap()
     {
          Text = "Keyhole Bitmap";

               // Load image.
          
          Image image = Image.FromFile(
               "..\\..\\..\\..\\Images and Bitmaps\\Apollo11FullColor.jpg");

               // Create clipping path.

          GraphicsPath path = new GraphicsPath();
          path.AddArc(80, 0, 80, 80, 45, -270);
          path.AddLine(70, 180, 170, 180);

               // Get size of clipping path.

          RectangleF rectf = path.GetBounds();
 
               // Create new bitmap initialized to transparent.

          bitmap = new Bitmap((int) rectf.Width, (int) rectf.Height, 
                              PixelFormat.Format32bppArgb);

               // Create Graphics object based on new bitmap.

          Graphics grfx = Graphics.FromImage(bitmap);

               // Draw original image on bitmap with clipping.

          grfx.SetClip(path);
          grfx.TranslateClip(-rectf.X, -rectf.Y);
          grfx.DrawImage(image, (int) -rectf.X, (int) -rectf.Y,
                                 image.Width, image.Height);
          grfx.Dispose();
     }
     protected override void DoPage(Graphics grfx, Color clr, int cx, int cy)
     {
          grfx.DrawImage(bitmap, (cx - bitmap.Width) / 2, 
                                 (cy - bitmap.Height) / 2,
                                 bitmap.Width, bitmap.Height);
     }
}
